### PR TITLE
[SPARK-30062][SQL] Add the IMMEDIATE statement to the DB2 dialect truncate implementation

### PR DESCRIPTION
**Pull request by ikarol**
### What changes were proposed in this pull request?
I've added a DB2 specific truncate implementation that adds an IMMEDIATE statement at the end of the query.

### Why are the changes needed?
I've encountered this issue myself while working with DB2 and trying to use truncate functionality.
A quick google search shows that some people have also encountered this issue before:
https://stackoverflow.com/questions/70027567/overwrite-mode-does-not-work-in-spark-sql-while-adding-data-in-db2
https://issues.apache.org/jira/browse/SPARK-30062

By looking into DB2 docs it becomes apparent that the IMMEDIATE statement is only optional if the table is column organized(though I'm not sure if it applies to all DB2 versions). So for the cases(such as mine) where the table is not column organized adding an IMMEDIATE statement becomes essential for the query to work.
https://www.ibm.com/support/knowledgecenter/en/SSEPGG_11.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0053474.html

Also, that might not be the best example, but I've found that DbVisualizer does add an IMMEDIATE statement at the end of the truncate command. Though, does it only for versions that are >=9.7
https://fossies.org/linux/dbvis/resources/profiles/db2.xml (please look at line number 473)

### Does this PR introduce _any_ user-facing change?
It should not, as even though the docs mention that if the TRUNCATE statement is executed in conjunction with IMMEDIATE, it has to be the first statement in the transaction, the JDBC connection that is established to execute the TRUNCATE statement has the auto-commit mode turned on. This means that there won't be any other query/statement executed prior within the same transaction.
https://www.ibm.com/docs/en/db2/11.5?topic=statements-truncate (see the description for IMMEDIATE)
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala#L49
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala#L57
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L108

### How was this patch tested?
Existing test case with slightly adjusted logic.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @huaxingao and @dilipbiswal FYI

**Comment by huaxingao** [comment]
I checked DB2 doc,  the IMMEDIATE clause is optional only in DB2 Version 11.5 Mod Pack 2 and later versions. I think it's reasonable to add IMMEDIATE in truncate statement to make it work with all versions of DB2. 

@ikarol Could you please add a test in `org.apache.spark.sql.jdbc.DB2IntegrationSuite` to test `df.write.mode(SaveMode.Overwrite).option("truncate", true)`?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala [hunk 1]
```diff
@@ -52,6 +52,15 @@ private object DB2Dialect extends JdbcDialect {
 
   override def isCascadingTruncateTable(): Option[Boolean] = Some(false)
 
+  // scalastyle:off line.size.limit
+  // See https://www.ibm.com/support/knowledgecenter/en/SSEPGG_11.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0053474.html
+  // scalastyle:on line.size.limit
+  override def getTruncateQuery(
+    table: String,
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:59]: nit: 4-space indentation

**Comment by huineng** [comment]
thanks for this , i applied the suggested code for 2 years now manually (recompiling spark) and i never had any issue. This merge would save me doing this manual compilation

**Comment by ikarol** [comment]
> I checked DB2 doc, the IMMEDIATE clause is optional only in DB2 Version 11.5 Mod Pack 2 and later versions. I think it's reasonable to add IMMEDIATE in truncate statement to make it work with all versions of DB2.
> 
> @ikarol Could you please add a test in `org.apache.spark.sql.jdbc.DB2IntegrationSuite` to test 
`df.write.mode(SaveMode.Overwrite).option("truncate", true)`?

@huaxingao 
Thank you for the review, I've made the changes you requested.
If there is still something missing, please let me know.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala [hunk 2]
```diff
@@ -198,4 +198,26 @@ class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
...
   }
+
+  test("SPARK-30062") {
+    val expectedResult = Set(
+      (42, "fred"),
+      (17, "dave")
+    ).map { case (x, y) =>
+      Row(Integer.valueOf(x), String.valueOf(y))
+    }
+    val df = sqlContext.read.jdbc(jdbcUrl, "tbl", new Properties).cache
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:209]: nit: I think we don't have to cache it.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala [hunk 3]
```diff
@@ -198,4 +198,26 @@ class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
...
+    df.write.mode(SaveMode.Append)
+      .jdbc(jdbcUrl, "tblcopy", new Properties)
+    df.write.mode(SaveMode.Append)
+      .jdbc(jdbcUrl, "tblcopy", new Properties)
+    df.write.mode(SaveMode.Append)
+      .jdbc(jdbcUrl, "tblcopy", new Properties)
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:215]: nit: put these in a loop?

**Review by huaxingao — APPROVED** [review]

**Comment by huaxingao** [comment]
Merged to master/3.2. 

@ikarol Thanks for your contribution and welcome to Spark community!

**Comment by huaxingao** [comment]
@ikarol is your jira id `ikarol`? I was trying to assign the jira to you, but somehow can't find your jira id.

**Comment by ikarol** [comment]
@huaxingao thanks for the warm welcome!

> @ikarol is your jira id `ikarol`? I was trying to assign the jira to you, but somehow can't find your jira id.

I did not have one when I've created this PR, but now I have and it's ikarol

**Comment by gatorsmile** [comment]
@huaxingao how about DB2 on z/OS?

**Comment by huaxingao** [comment]
@gatorsmile In z/OS DB2 IMMEDIATE clause is optional in v10, v11 and v12.

**Pull request closed without merging**